### PR TITLE
[MAISTRA-2096] Update Prometheus to 2.24 (OpenShift 4.7 version)

### PR DESCRIPTION
**Pull request by brian-avery**
I've updated maistra/prometheus/maistra-2.1 to match the OpenShift release-4.7 branch. This cherrypicks all of our Prometheus commits on top of that. In the process:

* XNS Informers has replaced the multilistwatcher code
* lots of dependencies have been updated in go.mod
* The SimonPasquier logging library that Prometheus uses is not compatible with the Kubernetes logger that Istio uses (they implement different interfaces). It's hacky, but the best solution I could find was to create an adapter to satisfy the interfaces. See logwrapper.go for more detail

To test this, I've deployed a 2.1 control plane with bookinfo, pushed traffic through and confirmed that I'm able to fetch metrics using the Prometheus UI. I then updated this to my image containing these changes and pushed more traffic through, again confirming that it shows up on the Prometheus UI

The multitenancy commit didn't apply cleanly (refactoring of discovery.go and manager.go), but switching to XNS informers helped. Other issues where go.mod had to be massaged. This picks up 2 years worth of Prometheus commits as well.

-----

Make test detects leaks in:
* klog-v2 -- known issue here https://github.com/kubernetes/client-go/issues/900 and here https://github.com/kubernetes/klog/issues/188
* go.opencensus.io -- known issue here https://github.com/census-instrumentation/opencensus-go/issues/1191

The go.census issue lists a workaround that I will add to our code. 

----
EndpointsDiscovery lists some failures that I'm looking into

### CODE REVIEW

File: discovery/kubernetes/kubernetes.go [hunk 1]
```diff
@@ -376,139 +395,64 @@ func (d *Discovery) Run(ctx context.Context, ch chan<- []*targetgroup.Group) {
...
+		eps := NewEndpoints(
+			log.With(d.logger, "role", "endpoint"),
+			svcInformer,
+			epInformer,
+			pInformer,
+		)
+		d.discoverers = append(d.discoverers, eps)
+		go eps.endpointsInf.Run(ctx.Done())
+		go eps.serviceInf.Run(ctx.Done())
+		go eps.podInf.Run(ctx.Done())
```
Comment (rcernich) [inline discovery/kubernetes/kubernetes.go]: As opposed to individual `Run()` invocations, I think we need to simply `Start()` the member roll controller.  I believe that invokes `Run()` for all informers.  (Should probably verify.)

**Comment by jwendell** [comment]
sorry, have to wait post submit

**Comment by brian-avery** [comment]
https://github.com/maistra/prometheus/pull/9/commits/9f7854704f5fc5a0cb179d90108afd20c3947796 

goleak.VerifyTestMain enables the Uber's go leak tester. Still investigating this, but when it is run with the Kubernetes tests, the endpoint tests time out. Disabling this causes the endpoint tests to pass.

File: discovery/kubernetes/kubernetes.go [hunk 2]
```diff
@@ -376,139 +395,64 @@ func (d *Discovery) Run(ctx context.Context, ch chan<- []*targetgroup.Group) {
 			go eps.podInf.Run(ctx.Done())
 		}
 	case RoleEndpoint:
-		for _, namespace := range namespaces {
-			e := d.client.CoreV1().Endpoints(namespace)
-			elw := &cache.ListWatch{
-				ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
-					options.FieldSelector = d.selectors.endpoints.field
```
Comment (dgn) [inline discovery/kubernetes/kubernetes.go:383]: I see you're removing the custom ListWatch's here- you can also use the informerFactory's `InformerFor()` method to get an informer, that'll allow you to keep the custom ListWatch funcs

File: .github/config.yml [hunk 3]
```diff
@@ -0,0 +1 @@
+COMMIT_MESSAGE_REGEX: "\bMAISTRA-\d+(?![.\d])"
```
Comment (jwendell) [inline .github/config.yml]: you can remove this file

File: .mergify.yml [hunk 4]
```diff
@@ -0,0 +1,31 @@
+pull_request_rules:
```
Comment (jwendell) [inline .mergify.yml]: you can remove this file

File: cmd/prometheus/logWrapper.go [hunk 5]
```diff
@@ -0,0 +1,100 @@
+// Copyright 2021 The Maistra Team
```
Comment (jwendell) [inline cmd/prometheus/logWrapper.go]: I think this should be `Red Hat, Inc.`

File: go.mod [hunk 6]
```diff
@@ -4,90 +4,98 @@ go 1.14
...
+	k8s.io/client-go v0.20.1
 	k8s.io/klog v1.0.0
-	k8s.io/klog/v2 v2.4.0
 )
 
-replace (
-	k8s.io/klog => github.com/simonpasquier/klog-gokit v0.3.0
-	k8s.io/klog/v2 => github.com/simonpasquier/klog-gokit/v2 v2.0.1
-)
+//BAVERY_TODO: fix before merging
```
Comment (jwendell) [inline go.mod]: remove this?

File: go.mod [hunk 7]
```diff
@@ -4,90 +4,98 @@ go 1.14
...
+	github.com/soheilhy/cmux v0.1.4
+	go.opencensus.io v0.22.0 // indirect
+	go.uber.org/zap v1.13.0 // indirect
+	//golang.org/x/sys v0.0.0-20191010194322-b09406accb47
+	golang.org/x/tools v0.0.0-20191029190741-b9c20aec41a5
 	google.golang.org/api v0.30.0
+	google.golang.org/api v0.8.0
+	google.golang.org/grpc v1.22.1
+	gopkg.in/yaml.v2 v2.2.2
+	//istio.io/pkg v0.0.0-20201230223204-2d0a1c8bd9e5
```
Comment (jwendell) [inline go.mod]: remove this?

File: util/testutil/testing.go [hunk 8]
```diff
@@ -24,17 +24,15 @@ package testutil
...
 	"testing"
-
-	"go.uber.org/goleak"
 )
 
 // TolerantVerifyLeak verifies go leaks but excludes the go routines that are
 // launched as side effects of some of our dependencies.
 func TolerantVerifyLeak(m *testing.M) {
-	goleak.VerifyTestMain(m,
+	/*goleak.VerifyTestMain(m,
```
Comment (jwendell) [inline util/testutil/testing.go]: do you want to add a comment, like pointing to an issue url?

File: discovery/kubernetes/kubernetes.go (hunk 1 above)
Comment (brian-avery) [reply discovery/kubernetes/kubernetes.go]: Done. Thanks.

File: discovery/kubernetes/kubernetes.go (hunk 2 above)
Comment (brian-avery) [reply discovery/kubernetes/kubernetes.go:383]: Thanks. All of the Prometheus tests as well as my own manual testing are passing right now. Lets merge for  now and we can optimize for that in the future. That would probably make future merges easier if the code changes.

**Review by jwendell — APPROVED** [review]

**Pull request merged**